src/pudl/workspace/datastore.py

```diff
@@ -154,106 +152,127 @@ def get_json_string(self) -> str:
         return json.dumps(self.datapackage_json, sort_keys=True, indent=4)
 
 
+class ZenodoDoiSettings(BaseSettings):
+    """Digital Object Identifiers pointing to currently used Zenodo archives."""
+
+    # Sandbox DOIs are provided for reference
```

Comment:
This works, and it makes sense to me to keep the sandbox DOIs around somewhere. It'd be nice if we could keep them somewhere that's actually encoded, as opposed to in a comment. But, we still want to be able to swap around "which datasets come from which environment" without accidentally committing a change that sends people to the sandbox randomly.

I think we sort of have these moving pieces at play, conceptually:

* there are multiple environments which use different DOIs for each dataset
* each dataset is drawn from one environment at a time, defaulting to prod
* the environment affects the HTTP access token as well as the base URL

Right now, the environment is encoded as "does the DOI include `10.5072` or `10.5281`?" or "does the URL include `sandbox`?". We could easily turn it into an explicit part of the data model, which will help us avoid the issue of committing sandbox changes completely. I'm thinking something like this:

```python
class ZenodoEnv(enum.Enum):
  SANDBOX: "sandbox"
  PRODUCTION: "production" 
  
class ZenodoEnvSettings(BaseModel):
  dois: ZenodoDoiSettings
  token: str
  base_url: HttpUrl
  
  def get_doi(self, dataset):
    return self.dois.__getattribute__(dataset)

class ZenodoFetcher:
  target_envs = ZenodoTargetEnvs.from_yaml("some_config.yaml")
  envs = {ZenodoEnv.SANDBOX: ZenodoEnvSettings(...), ZenodoEnv.PRODUCTION: ZenodoEnvSettings(...)}
  
  def _fetch_from_url(self, url, env):
    token = self.envs[env].token
    ...

  def get_resource(self, res: PudlResourceKey) -> bytes:
    env = self.target_envs[res.dataset]
    doi = self.envs[env].get_doi(res.dataset)
    ...
```

Do you think the benefits of the refactor (ZenodoFetcher not having to think about inter-env differences, lower chance of messing up the env config) are worth the time of actually doing it?

Reply:
I agree that this would be nice, but I'd prefer not to go all the way down the path of configuring the DOIs with environment variables and a config file right now. I'm doing this as an open source contribution because it's a low priority but it's annoying me. We change DOIs rarely, and just being able to cut-and-paste by hand a mix of sandbox and production DOIs will be a big ergonomic improvement. The sandbox DOIs also aren't particularly stable -- sometimes an update to the archivers necessitates starting a new archive lineage. So formally encoding them is of somewhat less utility than the production DOIs. And also every time we want to mess around with a new archive we're going to have to manually update DOIs anyway.

I think the right time to do this work might be when we have someone willing to pay for the (semi-) automation of creating fresh PRs from newly minted archives to attempt running through the ETL, since parsing / editing a YAML file that contains the DOIs will be cleaner and easier to automate.

Reply:
Sure, if we're mostly just In A Hurry this totally works :)